File: Python/Linked List/11_intersection_of_2_LL.py
"""
Idea:
Iterating the both LL in parallel, moving pointer of the LL with smaller value, & appending common values in 3rd LL
"""


# User function Template for python3


def findIntersection(head1, head2):
    p1 = head1
    p2 = head2
    head = None
    tail = None

    while p1 and p2:
        if p1.data > p2.data:
            # nodes don't match moving to next node in list with smaller node
            p2 = p2.next

        elif p2.data > p1.data:
            # nodes don't match moving to next node in list with smaller node
            p1 = p1.next

        else:
            # if nodes match:
            if head is None:
                head = tail = Node(p1.data)
            else:
                tail.next = Node(p1.data)
                tail = tail.next

            p1 = p1.next
            p2 = p2.next

    return head


# return head


# {
# Driver Code Starts
# Initial Template for Python 3

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class linkedList:
    def __init__(self):
        self.head = None
        self.tail = None

    def insert(self, data):
        if self.head is None:
            self.head = Node(data)
            self.tail = self.head
        else:
            self.tail.next = Node(data)
            self.tail = self.tail.next


def printList(head):
    while head:
        print(head.data, end=' ')
        head = head.next


if __name__ == '__main__':
    t = int(input())
    for _ in range(t):

        n1, n2 = [int(x) for x in input().split()]
        arr1 = [int(x) for x in input().split()]
        ll1 = linkedList()
        for i in arr1:
            ll1.insert(i)

        arr2 = [int(x) for x in input().split()]
        ll2 = linkedList()
        for i in arr2:
            ll2.insert(i)

        result = findIntersection(ll1.head, ll2.head)
        printList(result)
        print()

# } Driver Code Ends

File: C++/Linked List/11_intersection_of_2_LL.cpp
/*
Idea:
Iterating the both LL in parallel, moving pointer of the LL with smaller value, & appending common values in 3rd LL
*/

//{ Driver Code Starts
//

#include <bits/stdc++.h>
using namespace std;

struct Node
{
    int data;
    Node* next;
    Node(int val)
    {
        data = val;
        next = NULL;
    }
};

Node* inputList(int size)
{
    Node* head, * tail;
    int val;

    cin >> val;
    head = tail = new Node(val);

    while (--size)
    {
        cin >> val;
        tail->next = new Node(val);
        tail = tail->next;
    }

    return head;
}

void printList(Node* n)
{
    while (n)
    {
        cout << n->data << " ";
        n = n->next;
    }
}

Node* findIntersection(Node* head1, Node* head2);

int main()
{
    int t;
    cin >> t;
    while (t--)
    {
        int n, m;
        cin >> n >> m;

        Node* head1 = inputList(n);
        Node* head2 = inputList(m);

        Node* result = findIntersection(head1, head2);

        printList(result);
        cout << endl;
    }
    return 0;
}

// } Driver Code Ends



Node* findIntersection(Node* head1, Node* head2)
{
    Node* ans = new Node(-1), * ptr = ans;
    Node* t1 = head1, * t2 = head2;
    while (t1 != NULL && t2 != NULL)
    {
        if (t1->data == t2->data)
        { //if nodes match:
            Node* tp = new Node(t1->data);
            ptr->next = tp;
            ptr = ptr->next;
            t1 = t1->next;
            t2 = t2->next;
        }
        else if (t1->data < t2->data) // nodes don't match moving to next node in list with smaller node
            t1 = t1->next;
        else // nodes don't match moving to next node in list with smaller node 
            t2 = t2->next;
    }
    return ans->next;
}